src/imagetilesource.js

```diff
@@ -0,0 +1,274 @@
+         * @param {Number} level
+         * @param {OpenSeadragon.Point} point
+         */
+        getTileAtPoint: function (level, point) {
+            return new $.Point(0, 0);
+        },
+        /**
+         * This method is not implemented by this class other than to throw an Error
+         * announcing you have to implement it.  Because of the variety of tile
+         * server technologies, and various specifications for building image
+         * pyramids, this method is here to allow easy integration.
+         * @function
+         * @param {Number} level
+         * @param {Number} x
+         * @param {Number} y
+         * @throws {Error}
+         */
+        getTileUrl: function (level, x, y) {
+            var url = null;
+            if (level >= this.minLevel && level <= this.maxLevel) {
+                url = this.levels[level].url;
+            }
+            return url;
+        }
+    });
+
+    function buildLevels(image, minWidth, minHeight) {
+        var levels = [{
+                url: image.src,
+                width: image.naturalWidth,
+                height: image.naturalHeight
+            }];
+
+        var currentWidth = Math.floor(image.naturalWidth / 2);
+        var currentHeight = Math.floor(image.naturalHeight / 2);
+
+        if (currentWidth < minWidth || currentHeight < minHeight) {
+            return levels;
+        }
+
+        var bigCanvas = document.createElement("canvas");
+        var bigContext = bigCanvas.getContext("2d");
+
+        bigCanvas.width = currentWidth;
+        bigCanvas.height = currentHeight;
+        bigContext.drawImage(image, 0, 0, currentWidth, currentHeight);
+
+        if (isCanvasTainted(bigContext)) {
+            // If the canvas is tainted, we can't compute the pyramid.
+            return levels;
+        }
+        levels.splice(0, 0, {
+            url: bigCanvas.toDataURL(),
+            width: currentWidth,
+            height: currentHeight
+        });
+
+        var smallCanvas = document.createElement("canvas");
+        var smallContext = smallCanvas.getContext("2d");
+        while (currentWidth >= minWidth * 2 && currentHeight >= minHeight * 2) {
+            currentWidth = Math.floor(currentWidth / 2);
+            currentHeight = Math.floor(currentHeight / 2);
+            smallCanvas.width = currentWidth;
+            smallCanvas.height = currentHeight;
+            smallContext.drawImage(bigCanvas, 0, 0, currentWidth, currentHeight);
+
+            levels.splice(0, 0, {
+                url: smallCanvas.toDataURL(),
+                width: currentWidth,
+                height: currentHeight
+            });
+
+            var tempCanvas = bigCanvas;
+            bigCanvas = smallCanvas;
+            smallCanvas = tempCanvas;
+
+            var tempContext = bigContext;
+            bigContext = smallContext;
+            smallContext = tempContext;
+        }
+        return levels;
+    }
+
+    function isCanvasTainted(context) {
```

Comment:
Since this is a general purpose function, I suppose it could go on the OpenSeadragon object (in case we need to use it elsewhere in the future).